File: Programmers/LV1_개인정보_수집기간.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>

using namespace std;

// 9:45 - 10:09

int days(string date) {
    int yy = stoi(date.substr(0, 4));
    int mm = stoi(date.substr(5, 7));
    int dd = stoi(date.substr(8, 10));
    
    return (yy - 2000) * 336 + mm * 28 + dd;
}

vector<int> solution(string today, vector<string> terms, vector<string> privacies) {
    vector<int> answer;
    map<string, int> termsm;
    int today_days = days(today);
    
    // 약관 : 유효기간일수 map
    for(string term : terms) {
        string yack = term.substr(0, term.find(' '));
        int days = 28 * stoi(term.substr(term.find(' ') + 1));
        termsm[yack] = days;
    }
    
    // 가입일수 : 약관 확인
    for(int i = 0; i < privacies.size(); i++) {
        string privacy = privacies[i];
        int entered_days = days(privacy.substr(0, privacy.find(' ')));
        string yack_privacy = privacy.substr(privacy.find(' ') + 1);
        if(entered_days + termsm[yack_privacy] <= today_days)  answer.push_back(i + 1);        
    }
    
    return answer;
}

File: Programmers/LV1_개인정보_수집기간.java
import java.util.HashMap;
import java.util.Map;
import java.util.List;
import java.util.ArrayList;

class Solution {
    
    public int[] solution(String today, String[] terms, String[] privacies) {
        // int[] answer = {};
        List<Integer> answer = new ArrayList<>();
        int today_days = days(today);
        Map<String, Integer> termMap = new HashMap<>();
        
        // 약관-일수 매핑
        for(String s : terms) {
            String[] term = s.split(" ");
            termMap.put(term[0], Integer.parseInt(term[1]) * 28);
        }
        
        for(int i = 0; i < privacies.length; i++) {
            String[] privacy = privacies[i].split(" ");
            
            if(days(privacy[0]) + termMap.get(privacy[1]) <= today_days)
                answer.add(i + 1);
        }
        
        return answer.stream().mapToInt(integer -> integer).toArray();
    }
    
    public int days(String input) {
        String[] date = input.split("\\.");
        int yy = Integer.parseInt(date[0]);
        int mm = Integer.parseInt(date[1]);
        int dd = Integer.parseInt(date[2]);
        
        return (yy - 2000) * 12 * 28 + mm * 28 + dd;
    }
}